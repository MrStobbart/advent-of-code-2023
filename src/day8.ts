type Node2 = { name: string; L: string; R: string };

type Direction = "L" | "R";
type Node = Record<Direction, string> & { name: string };

type Nodes = Record<string, Node>;

type ParsedInput = {
  directions: Direction[];
  nodes: Nodes;
  startNodeName: string;
};

export const parseNode = (rawNode: string): Node => {
  if (rawNode.length !== 16) {
    throw new Error(`rawNode ${rawNode} does not have a length of 16 chars`);
  }

  const name = rawNode.substring(0, 3);
  const left = rawNode.substring(7, 10);
  const right = rawNode.substring(12, 15);
  return { name, R: right, L: left };
};

export const parseInput = (input: string): ParsedInput => {
  const directions = input.split("\n\n")[0]?.split("")! as Direction[];
  const rawNodes = input.split("\n\n")[1]!.split("\n");
  let startNodeName = "";
  const nodes = Object.fromEntries(
    rawNodes.map((rawNode, index) => {
      const node = parseNode(rawNode);
      if (index === 0) {
        startNodeName = node.name;
      }
      return [node.name, node];
    })
  );

  return { directions, nodes, startNodeName };
};

export const navigateInput = ({ directions, nodes }: ParsedInput): number => {
  const startNodeName = "AAA";
  const destination = "ZZZ";
  if (nodes[startNodeName] === undefined || nodes[destination] === undefined) {
    console.error("Start or destination not found", {
      start: startNodeName,
      destination,
      nodes,
    });

    return 0;
  }

  let currentNodeName = startNodeName;
  let steps = 0;

  while (currentNodeName !== destination) {
    if (steps > 1000000) {
      throw new Error(`Abort, this is already to high`);
    }
    const direction = directions[steps % directions.length]!;
    const node = nodes[currentNodeName];
    if (node === undefined) {
      throw new Error(`Node ${currentNodeName} does not exist`);
    }

    currentNodeName = node[direction];
    steps++;
  }

  return steps;
};
