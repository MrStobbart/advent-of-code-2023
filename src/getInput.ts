export const getInputForDay = (day: number) => {
  const fileName = `./data/day${day}.input.txt`;
  try {
    const file = Bun.file(fileName);
    return file.text();
  } catch (error) {
    throw new Error(
      `Could not load file with name ${fileName}`,
      error as Error
    );
  }
};
