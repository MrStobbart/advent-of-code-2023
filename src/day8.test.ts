import { describe, expect, it } from "bun:test";
import { navigateInput, parseInput, parseNode } from "./day8";
import { getInputForDay } from "./getInput";

const testInput1 = `RL

AAA = (BBB, CCC)
BBB = (DDD, EEE)
CCC = (ZZZ, GGG)
DDD = (DDD, DDD)
EEE = (EEE, EEE)
GGG = (GGG, GGG)
ZZZ = (ZZZ, ZZZ)`;

const testInput2 = `LLR

AAA = (BBB, BBB)
BBB = (AAA, ZZZ)
ZZZ = (ZZZ, ZZZ)`;

const input = await getInputForDay(8);

describe("Part1", () => {
  it("parses a node correctly", () => {
    expect(parseNode("AAA = (BBB, CCC)")).toEqual({
      name: "AAA",
      L: "BBB",
      R: "CCC",
    });
  });

  it("parses the testinput1 correctly", () => {
    expect(parseInput(testInput1)).toMatchSnapshot();
  });

  it("navigates testInput1 correctly", () => {
    const parsedInput = parseInput(testInput1);
    expect(navigateInput(parsedInput)).toBe(2);
  });

  it("navigates testInput2 correctly", () => {
    const parsedInput = parseInput(testInput2);
    expect(navigateInput(parsedInput)).toBe(6);
  });

  it("navigates the input correctly", () => {
    const parsedInput = parseInput(input);
    expect(navigateInput(parsedInput)).toBe(6);
  });
});
