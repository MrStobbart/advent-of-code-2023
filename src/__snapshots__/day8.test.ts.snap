// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Part1 parses the testinput1 correctly 1`] = `
{
  "directions": [
    "R",
    "L",
  ],
  "nodes": {
    "AAA": {
      "L": "BBB",
      "R": "CCC",
      "name": "AAA",
    },
    "BBB": {
      "L": "DDD",
      "R": "EEE",
      "name": "BBB",
    },
    "CCC": {
      "L": "ZZZ",
      "R": "GGG",
      "name": "CCC",
    },
    "DDD": {
      "L": "DDD",
      "R": "DDD",
      "name": "DDD",
    },
    "EEE": {
      "L": "EEE",
      "R": "EEE",
      "name": "EEE",
    },
    "GGG": {
      "L": "GGG",
      "R": "GGG",
      "name": "GGG",
    },
    "ZZZ": {
      "L": "ZZZ",
      "R": "ZZZ",
      "name": "ZZZ",
    },
  },
  "startNodeName": "AAA",
}
`;
